Lib/asyncio/pools.py

```diff
@@ -0,0 +1,173 @@
+
+
+class ThreadPool(AbstractPool):
+    """Asynchronous threadpool for running IO-bound functions.
+
+    Directly calling an IO-bound function within the main thread will block
+    other operations from occurring until it is completed. By using a
+    threadpool, several IO-bound functions can be ran concurrently within
+    their own threads, without blocking other operations.
+
+    The optional argument *concurrency* sets the number of threads within the
+    threadpool. If *concurrency* is `None`, the maximum number of threads will
+    be used; based on the number of CPU cores.
+
+    This threadpool is intended to be used as an asynchronous context manager,
+    using the `async with` syntax, which provides automatic initialization and
+    finalization of resources. For example:
+
+    import asyncio
+
+    def blocking_io():
+        print("start blocking_io")
+        with open('/dev/urandom', 'rb') as f:
+            f.read(100_000)
+        print("blocking_io complete")
+
+    def other_blocking_io():
+        print("start other_blocking_io")
+        with open('/dev/zero', 'rb') as f:
+            f.read(10)
+        print("other_blocking_io complete")
+
+    async def main():
+        async with asyncio.ThreadPool() as pool:
+            await asyncio.gather(
+                pool.run(blocking_io),
+                pool.run(other_blocking_io))
+
+    asyncio.run(main())
+    """
+
+    def __init__(self, concurrency=None):
+        if concurrency is None:
+            concurrency = min(32, (os.cpu_count() or 1) + 4)
+
+        self._concurrency = concurrency
+        self._running = False
+        self._closed = False
+        self._loop = None
+        self._pool = None
+
+    async def astart(self):
+        self._loop = events.get_running_loop()
+        await self._spawn_threadpool()
+
+    async def __aenter__(self):
+        await self.astart()
+        return self
+
+    async def aclose(self):
+        await self._shutdown_threadpool()
+
+    async def __aexit__(self, exc_type, exc_value, exc_traceback):
+        await self.aclose()
+
+    async def run(self, /, func, *args, **kwargs):
+        if not self._running:
+            raise RuntimeError(f"unable to run {func!r}, "
+                               "threadpool is not running")
+
+        func_call = functools.partial(func, *args, **kwargs)
+        executor = self._pool
+        return await futures.wrap_future(
+            executor.submit(func_call), loop=self._loop)
+
+    async def _spawn_threadpool(self):
+        """Schedule the spawning of the threadpool.
+
+        Asynchronously spawns a threadpool with *concurrency* threads.
+        """
+        if self._running:
+            raise RuntimeError("threadpool is already running")
+
+        if self._closed:
+            raise RuntimeError("threadpool is closed")
+
+        future = self._loop.create_future()
+        thread = threading.Thread(target=self._do_spawn, args=(future,))
```

Comment:
I'm not sure we need this complication, why don't we call `concurrent.futures.ThreadPoolExecutor()` here directly? Yes, it'll block for a slightly longer amount of time, but IMO that's totally OK. The implementation would be simpler.

Reply:
@1st1 
> I'm not sure we need this complication, why don't we call concurrent.futures.ThreadPoolExecutor() here directly? Yes, it'll block for a slightly longer amount of time, but IMO that's totally OK.

The current implementation for `astart()` and `aclose()` is based on the conversation in the bpo issue, mainly starting from [here](https://bugs.python.org/issue32309#msg355862):

> Thus btw makes me think that initializing threads/processes in `__init__` is a bad idea, as spawning can be asynchronous.

In the case of `astart()`, I don't think making it block would be a huge deal in most situations, as the main bottleneck would be waiting for the threads to spawn. This could be a concern when using a large number of threads (depending on the eventual native implementation of `asyncio.ThreadPool`), but otherwise I don't think it would be a real concern for the majority of users. Especially since `ThreadPoolExecutor` spawns threads on-demand as needed, rather than all of them at startup.

In the case of `aclose()`, I think that calling `executor.shutdown()` by itself would be rather detrimental, especially when using `wait=True` (which is the default). This would mean that the event loop would be effectively blocked while the executor is waiting on pending futures to be finished, no?

Instead, with the current implementation, it would allow the coroutine calling `await aclose()` (or exiting the `async with` scope) to suspend while the executor is shutting down in a separate thread until the awaited future is done, allowing tasks or other coroutines to complete work in the meantime. This could be quite impactful to performance, and would allow for more effective multi-tasking.

Thus, I think we could reasonable change `astart()` to directly initialize `ThreadPoolExecutor()` for simplification, but I strongly think that the current `aclose()` implementation should remain for the above reasons.

Reply:
Ah, you're right. Let's keep the current implementation for both `astart` and `aclose`

Reply:
How about using `loop.run_in_executor` instead of creating new thread only for starting/closing new threadpool?

We can assume loop.default_executor is ThreadPoolExecutor because:

```
    def set_default_executor(self, executor):
        if not isinstance(executor, concurrent.futures.ThreadPoolExecutor):
            warnings.warn(
                'Using the default executor that is not an instance of '
                'ThreadPoolExecutor is deprecated and will be prohibited '
                'in Python 3.9',
                DeprecationWarning, 2)
        self._default_executor = executor
```

Reply:
> How about using loop.run_in_executor instead of creating new thread only for starting/closing new threadpool?

+1.

Reply:
That's a great idea @methane, thanks for the suggestion.